### PR TITLE
[63] Assassin -> Resurrect as Police

### PR DESCRIPTION
**Pull request by thenewtonium**
This builds on PR #59 to address issue #63.

It adds a new `Assassin -> Resurrect as Police` export to ~~`CorePlugin`~~ (EDIT: moved to `PolicePlugin`). The user selects from all *dead, non-police, non-hidden* assassins, and is prompted to give a new initial pseudonym for the selected assassin. The assassin is then cloned (with a new identifier), with the clone set to be police and have the entered pseudonym as its sole pseudonym.

### CODE REVIEW

File: AU2/plugins/CorePlugin.py [hunk 1]
```diff
@@ -381,6 +406,27 @@ def answer_core_plugin_update_assassin(self, html_response_args: Dict):
             return_components += p.on_assassin_update(assassin, html_response_args)
         return return_components
 
+    def gather_dead_non_police(self) -> List[str]:
+        death_manager = DeathManager(perma_death=True)
```
Comment (jsyiek) [inline AU2/plugins/CorePlugin.py:410]: I see why you've done it this way, but `perma_death=True` is incorrect behavior for May Week games. This is fine in some plugins (e.g., Competency and Targeting) that explicitly are not used for May Week. How about setting `perma_death=True` iff the TargetingPlugins is enabled?

File: AU2/plugins/CorePlugin.py [hunk 2]
```diff
@@ -350,8 +377,6 @@ def answer_suppress_exports(self, htmlResponse):
 
     def ask_core_plugin_create_assassin(self):
         components = []
-        for p in PLUGINS:
```
Comment (jsyiek) [inline AU2/plugins/CorePlugin.py:353]: Accidental delete?

File: AU2/database/model/Assassin.py [hunk 3]
```diff
@@ -173,6 +195,9 @@ def pseudonyms_until(self, ts: dt.datetime = get_now_dt()) -> Iterator[str]:
         Yields:
             the next pseudonym of the assassin that is valid before datetime `ts`.
         """
+        if not ts:
```
Comment (jsyiek) [inline AU2/database/model/Assassin.py:198]: We're checking for None here, so it's better to be explicit so that devs inputting non-`datetime` data by mistake don't accidentally get `dt.now()`

```suggestion
        if ts is None:
```

**Review by jsyiek — CHANGES_REQUESTED** [review]
Initial review. I'm not sure what's from your previous PR and what's from this one - could you rebase master back into this branch to pick up changes?

File: AU2/plugins/CorePlugin.py (hunk 1 above)
Comment (thenewtonium) [reply AU2/plugins/CorePlugin.py:410]: Ah that's a good point.

It feels kind of hacky for `CorePlugin` to be checking for a specific plugin, but a better solution doesn't come to mind...

Comment (thenewtonium) [reply AU2/plugins/CorePlugin.py]: Maybe since `perma_death=False` just makes the list empty (so effectively disables the export) the export itself should be part of `TargetingPlugin`?

File: AU2/plugins/CorePlugin.py [hunk 4]
```diff
@@ -216,7 +225,6 @@ def on_assassin_request_update(self, assassin: Assassin):
             InputWithDropDown(self.html_ids["Water Status"], "Water Status", WATER_STATUSES, selected=assassin.water_status),
             InputWithDropDown(self.html_ids["College"], "College", COLLEGES, selected=assassin.college),
             LargeTextEntry(self.html_ids["Notes"], "Notes", default=assassin.notes),
-            #Checkbox(self.html_ids["Police"], "Police? (y/n)", checked=assassin.is_police)
```
Comment (NerdyNinja11) [inline AU2/plugins/CorePlugin.py:219]: Maybe leave another comment explaining why this is commented, or just remove the comment altogether. Not obvious why it's there otherwise

File: AU2/plugins/CorePlugin.py [hunk 5]
```diff
@@ -226,13 +234,12 @@ def on_assassin_update(self, assassin: Assassin, htmlResponse: Dict) -> List[HTM
...
         [assassin.edit_pseudonym(i, v.text, v.valid_from) for i, v in pseudonym_updates.edited.items()]
         [assassin.delete_pseudonym(i) for i in pseudonym_updates.deleted_indices]
-
+        # set other attributes
         assassin.real_name = htmlResponse[self.html_ids["Real Name"]]
         assassin.pronouns = htmlResponse[self.html_ids["Pronouns"]]
         assassin.address = htmlResponse[self.html_ids["Address"]]
         assassin.college = htmlResponse[self.html_ids["College"]]
         assassin.notes = htmlResponse[self.html_ids["Notes"]]
-        #assassin.is_police = htmlResponse[self.html_ids["Police"]]
```
Comment (NerdyNinja11) [inline AU2/plugins/CorePlugin.py:235]: (same as above) Maybe leave another comment explaining why this is commented, or just remove the comment altogether. Not obvious why it's there otherwise

**Review by NerdyNinja11 — COMMENTED** [review]
Not sure how useful it would be, but since we're in the business now of adding tests, you could add a test for this?

File: AU2/plugins/CorePlugin.py (hunk 4 above)
Comment (thenewtonium) [reply AU2/plugins/CorePlugin.py:219]: (The change here *is* removing the comment (that's what the - and the red mean).)

Comment (NerdyNinja11) [reply AU2/plugins/CorePlugin.py:219]: Oh oops

File: AU2/plugins/CorePlugin.py [hunk 6]
```diff
@@ -401,6 +408,29 @@ def answer_core_plugin_update_assassin(self, html_response_args: Dict):
             return_components += p.on_assassin_update(assassin, html_response_args)
         return return_components
 
+    def gather_dead_non_police(self) -> List[str]:
+        # use targeting plugin as proxy for whether we have perma-death
```
[the remark this answers is not in the document]
Comment (NerdyNinja11) [reply AU2/plugins/CorePlugin.py:412]: There are a few edge cases if the umpire decides to run their game in an unusual way, but I think that would work.
Using CompetencyPlugin seems fine though (although I do agree that it seems kinda hacky)

File: AU2/test/plugins/util/custom_plugins/test_PolicePlugin.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        # check that the list of players that can be resurrected is correct
+        # (only the dead non-police, non-hidden assassins should be shown!)
+        assert set(plugin.gather_dead_non_police()) == {p[i] + " identifier" for i in range(k, 2*k-1)}
+
+        # check that players get resurrected correctly
+        ##############################################
+        new_pseudonym = "New Pseudonym"
+        old_assassin = ASSASSINS_DATABASE.get(p[k] + " identifier")
+        old_idents = set(ASSASSINS_DATABASE.get_identifiers(include_hidden=lambda _: True))
+        # the way identifiers are generated in test mode messes with cloning
```
Comment (NerdyNinja11) [inline AU2/test/plugins/util/custom_plugins/test_PolicePlugin.py:43]: Since this has caused issues for both of us, maybe we need to rewrite how test mode treats assassins a bit.

**Review by NerdyNinja11 — APPROVED** [review]
lgtm

File: AU2/test/plugins/util/custom_plugins/test_PolicePlugin.py (hunk 7 above)
Comment (thenewtonium) [reply AU2/test/plugins/util/custom_plugins/test_PolicePlugin.py:43]: Yeah...
In this particular case the annoying thing is that if identifiers worked normally in test mode all the other tests would need to be rewritten. Luckily we don't have that many tests at the moment...
